ipykernel/kernelapp.py

```diff
@@ -446,6 +447,18 @@ def init_kernel(self):
                                 profile_dir=self.profile_dir,
                                 user_ns=self.user_ns,
         )
+        kernel.stdin_stream = stdin_stream
+
+        def handle_msg(msg):
+            idents, msg = self.session.feed_identities(msg, copy=False)
+            try:
+                msg = self.session.deserialize(msg, content=True, copy=False)
+            except:
+                self.log.error("Invalid Message", exc_info=True)
+                return
+            kernel._stdin_msg = msg
+
+        kernel.stdin_stream.on_recv(handle_msg, copy=False)
```

Comment:
I think all these new lines here should be moved to the `Kernel` class (present in `kernelbase.py`). In particular, `handle_msg` should be part of `dispatch_stdin` (following what `dispatch_control` and `dispatch_shell` do for `control_stream` and `shell_streams`, respectively).

However, I don't know what the implications of adding a new stream here are for the Jupyter protocol. Pinging @takluyver for his insights about that.

Reply:
I moved it. I don't want to process the messages recieved on stdin so I don't think `dispatch_stdin` is a good idea. We are not dispatching the messages, only saving the most recent. A stream is just a wrapper around a socket. I am not adding a socket, so the protocol is unchanged. The only reason I need a stream around the socket is to get a qt event when a message is recieved.

Reply:
Yes, AFAIK @impact27 is right - there shouldn't be a difference to the protocol. A pyzmq stream is a wrapper around a socket to tie it to an event loop. The stdin socket already existed, so whether the code wraps it in a stream or not should be an implementation detail.

(I haven't tried to review the PR, just answering the question I was mentioned for)